### PR TITLE
Adding ability to add custom providers that add your own variables to config freemarker template.

### PR DESCRIPTION
**Pull request by allippert**

### CODE REVIEW

**Comment by tkrille** [comment]
Hi @allippert,

thanks for your PR and your interest in Dropwizard Template Config. I will have a look at this in the evening.

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigBundleConfiguration.java [hunk 1]
```diff
@@ -64,6 +67,11 @@ public TemplateConfigBundleConfiguration charset(Charset charset) {
     }
 
     /**
+     * Get the set of custom providers used to add variables to the configuration template (Default: Empty Set)
+     */
+    public Set<TemplateConfigProvider> customProviders() { return customProviders; }
```
Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigBundleConfiguration.java]: Please, put this in 3 separate lines.

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigProvider.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package de.thomaskrille.dropwizard_template_config;
+
+import java.util.Map;
+
+public interface TemplateConfigProvider {
+
+    /**
+     * @return The namespace to use when accessing objects in the config template.  Objects can either be accessed
```
Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigProvider.java]: Typo: 2 spaces between `config template.` and `Objects can either`.

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigProvider.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+     * in the template with ${namespace.object}
```
Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigProvider.java]: I don't think this sentence is complete :)

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigProvider.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+     */
+    String getNamespace();
+
+    /**
+     * @return A data model that the freemarker engine will use when parsing the config template.
+     */
+    Map<String, String> getDataModel();
```
Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigProvider.java]: I would rather call this method `getVariables`.

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigProvider.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package de.thomaskrille.dropwizard_template_config;
+
+import java.util.Map;
+
+public interface TemplateConfigProvider {
```
Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigProvider.java]: I would rather call this interface `TemplateConfigVariablesProvider`.

File: src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigBundleConfiguration.java [hunk 6]
```diff
@@ -125,4 +133,12 @@ public TemplateConfigBundleConfiguration outputPath(String outputPath) {
         this.outputPath = Optional.of(outputPath);
         return this;
     }
+
+    /**
+     * Add a custom provider used to add your own variables to the configuration template.
+     */
+    public TemplateConfigBundleConfiguration withCustomProvider(TemplateConfigProvider customProvider) {
```
Comment (tkrille) [inline src/main/java/de/thomaskrille/dropwizard_template_config/TemplateConfigBundleConfiguration.java]: I think it's better to call this method `addCustomProvider`. This way, you instantly see that the method adds a new provider. The prefix `with` does not carry this information.

**Comment by tkrille** [comment]
Hi @allippert,

this looks really good to me. Thank you very much for this feature! Please, fix the formatting issues and think about the naming suggestions.

**Comment by allippert** [comment]
Hi @tkrille - I just updated this based on your suggestions.  Thanks.

**Comment by tkrille** [comment]
Thanks again, @allippert, for the new feature. I will prepare a release during the next days. I want to do a little refactoring beforehand, as passing the configuration to `TemplateConfigurationSourceProvider` got a little out of hand :) I do release announcements via Twitter. Do you have a Twitter account, that I can mention in the release announcement?

**Pull request merged**